modules/git/utils_closererror.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package git
+
+import (
+	"io"
+	"sync"
+)
+
+type closerError struct {
+	io.Closer
+	err  error
+	lock sync.Mutex
+}
+
+func (c *closerError) CloseWithError(err error) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.err != nil {
+		return c.err
+	}
+	if c.Closer != nil {
+		_ = c.Closer.Close()
+	}
+	if err == nil {
+		err = io.ErrClosedPipe
+	}
```

Comment:
It breaks the behavior of `os.PipeWriter`, the behavior is:

```
// CloseWithError closes the writer; subsequent reads from the
// read half of the pipe will return no bytes and the error err,
// or EOF if err is nil.
//
// CloseWithError never overwrites the previous error if it exists
// and always returns nil.
func (w *PipeWriter) CloseWithError(err error) error {
	return w.p.closeWrite(err)
}

func (p *pipe) closeWrite(err error) error {
	if err == nil {
		err = EOF
	}
	p.werr.Store(err)
	p.once.Do(func() { close(p.done) })
	return nil
}
```

Reply:
Hmm... I'm just gonna revert this - we never looked for ErrClosedPipe or its string error type and I'm not overloading the Close() to call CloseWithError().